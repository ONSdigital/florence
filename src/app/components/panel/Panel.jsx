import React from "react";
import PropTypes from "prop-types";

// Types: warning, error, announcement, information, success, bare
const propTypes = {
    type: PropTypes.string.isRequired,
    body: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    heading: PropTypes.string,
    bannerHeading: PropTypes.bool,
    icon: PropTypes.oneOf(["exclamation", "tick", "arrow"]),
    input: PropTypes.shape({
        label: PropTypes.string,
        type: PropTypes.string,
        id: PropTypes.string,
        pattern: PropTypes.string,
        inputMode: PropTypes.string,
        required: PropTypes.bool
    })
};
const Panel = props => {
    let panelClass = `panel`;
    panelClass += ` panel--${props.type}`;
    panelClass += props.bannerHeading ? `` : ` panel--no-title`;

    const headingSection = (
        <div className={"panel__header"}>
            <h2 id="error-summary-title" data-qa="header" className="panel__title">
                {props.heading}
            </h2>
        </div>
    );
    const tickSVG = (
        <svg className="svg-icon" viewBox="0 0 13 10" xmlns="http://www.w3.org/2000/svg" focusable="false">
            <path
                d="M14.35,3.9l-.71-.71a.5.5,0,0,0-.71,0h0L5.79,10.34,3.07,7.61a.51.51,0,0,0-.71,0l-.71.71a.51.51,0,0,0,0,.71l3.78,3.78a.5.5,0,0,0,.71,0h0L14.35,4.6A.5.5,0,0,0,14.35,3.9Z"
                transform="translate(-1.51 -3.04)"
            />
        </svg>
    );
    const arrowSVG = (
        <svg xmlns="http://www.w3.org/2000/svg" width="24px" height="24px" viewBox="0 0 24 24" focusable="false" aria-hidden="true">
            <path
                d="M4.2,12c0-0.6,0.4-1,1-1h11.2l-4.9-4.9c-0.4-0.4-0.4-1,0-1.4c0.4-0.4,1-0.4,1.4,0l6.6,6.6c0.4,0.4,0.4,1,0,1.4l-6.6,6.6c-0.4,0.4-1,0.4-1.4,0c-0.4-0.4-0.4-1,0-1.4l4.9-4.9H5.2C4.7,13,4.2,12.6,4.2,12z"
                fill="currentColor"
            />
        </svg>
    );

    let iconSection;
    switch (props.icon) {
        case "tick":
            iconSection = <span className="panel__icon">{tickSVG}</span>;
            break;
        case "arrow":
            iconSection = (
                <span className="panel__icon" aria-hidden="true">
                    {arrowSVG}
                </span>
            );
            break;
        case "exclamation":
            iconSection = (
                <span className="panel__icon" aria-hidden="true">
                    !
                </span>
            );
            break;
    }
    let bodySection = (
        <div className="panel__body">
            {props.heading && !props.bannerHeading && (
                <p className={"panel__" + props.type}>
                    <strong className={"font-weight--600"}>{props.heading}</strong>
                </p>
            )}
            {props.body}
        </div>
    );
    let inputSection;
    if (props.input != null) {
        inputSection = (
            <div className="field">
                <label className="label  " htmlFor="number">
                    {props.input.label}
                </label>
                <input
                    type={props.input.type}
                    id={props.input.id}
                    className={`input input--text input-type__input  input--props.type`}
                    pattern={props.input.pattern}
                    inputMode={props.input.inputMode}
                    required={props.input.required}
                />
            </div>
        );
    }

    //______________________________________________________________________
    //{props.heading && props.bannerHeading && headingSection}
    //                     {props.input && inputSection}
    //______________________________________________________________________
    return (
        <div className={props.type}>
            <div className={props.type === "announcement" ? "padding-right--1 padding-left--1" : ""}>
                <div className={panelClass}>
                    {props.heading && props.bannerHeading && headingSection}
                    {props.icon && iconSection}
                    <span className="visually-hidden">{props.type}</span>
                    {props.body && bodySection}
                    {props.input && inputSection}
                </div>
            </div>
        </div>
    );
};
Panel.propTypes = propTypes;
export default Panel;
