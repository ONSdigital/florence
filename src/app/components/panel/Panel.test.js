import React from "react";
import Panel from "./Panel.jsx";
import renderer from "react-test-renderer";

test("error panel with a heading and a bullet point list ", () => {
    const ErrorPanelBody = () => {
        return (
            <ol className="list">
                <li className="list__item ">
                    <a href="#container-test-number" className="list__link js-inpagelink">
                        Enter a number
                    </a>
                </li>
                <li className="list__item ">
                    <a href="#container-test-percent" className="list__link js-inpagelink">
                        Enter a number less than or equal to 100
                    </a>
                </li>
            </ol>
        );
    };

    const props = {
        type: "error",
        heading: "There are 2 problems: ",
        body: <ErrorPanelBody />,
        bannerHeading: true,
    };
    const component = renderer.create(<Panel {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
});

test("information panel with no heading and raw string for body", () => {
    const props = {
        type: "information",
        body:
            "In principle but likewise practice this is a relatively simple but important message. " +
            "The fact of the matter is that you had absolutely no idea and therefore did not know " +
            "whether or not it was vital for you to spend the time to read this message until you actually " +
            "started to consume the words which this message contains. Now that you have " +
            "started to read it this message can help you realise that the nature of this message is to " +
            "inform you that you have indeed actually now read this message.",
    };
    const component = renderer.create(<Panel {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
});

test("error type panel with a regular heading and body", () => {
    const props = {
        type: "error",
        heading: "Fix the following: ",
        body: "Email address or password are incorrect",
    };
    const component = renderer.create(<Panel {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
});

test("success panel with a tick icon and a raw text body", () => {
    const props = {
        type: "success",
        body: "Information has been successfully submitted",
        icon: "tick",
    };
    const component = renderer.create(<Panel {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
});

test("error panel with a regular heading and an input field", () => {
    const props = {
        type: "error",
        heading: "Enter a number ",
        bannerHeading: false,
        input: {
            pattern: "[0-9]",
            id: "number",
            type: "text",
            label: "Number of employees paid monthly",
            inputMode: "numeric",
        },
    };
    const component = renderer.create(<Panel {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
});
test("announcement panel with custom html in body and an icon and no heading", () => {
    const AnnouncementPanelBody = () => {
        return (
            <p>
                "Coronavirus lockdown: stay at home Coronavirus (COVID-19) remains a serious threat across the country. <a href="#0">Find out more</a>
            </p>
        );
    };
    const props = {
        type: "announcement",
        body: <AnnouncementPanelBody />,
        icon: "arrow",
    };
    const component = renderer.create(<Panel {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
});
test("warning panel with an exclamation icon and raw string body", () => {
    const props = {
        type: "warning",
        body: "All the information about this person will be deleted",
        icon: "exclamation",
    };
    const component = renderer.create(<Panel {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
});
