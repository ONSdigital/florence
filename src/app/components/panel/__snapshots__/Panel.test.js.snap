// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`announcement panel with custom html in body and an icon and no heading 1`] = `
<div
  className="announcement"
>
  <div
    className="padding-right--1 padding-left--1"
  >
    <div
      className="panel panel--announcement panel--no-title"
    >
      <span
        aria-hidden="true"
        className="panel__icon"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          height="24px"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.2,12c0-0.6,0.4-1,1-1h11.2l-4.9-4.9c-0.4-0.4-0.4-1,0-1.4c0.4-0.4,1-0.4,1.4,0l6.6,6.6c0.4,0.4,0.4,1,0,1.4l-6.6,6.6c-0.4,0.4-1,0.4-1.4,0c-0.4-0.4-0.4-1,0-1.4l4.9-4.9H5.2C4.7,13,4.2,12.6,4.2,12z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        className="visually-hidden"
      >
        announcement
      </span>
      <div
        className="panel__body"
      >
        <p>
          "Coronavirus lockdown: stay at home Coronavirus (COVID-19) remains a serious threat across the country. 
          <a
            href="#0"
          >
            Find out more
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`error panel with a heading and a bullet point list  1`] = `
<div
  className="error"
>
  <div
    className=""
  >
    <div
      className="panel panel--error"
    >
      <div
        className="panel__header"
      >
        <h2
          className="panel__title"
          data-qa="header"
          id="error-summary-title"
        >
          There are 2 problems: 
        </h2>
      </div>
      <span
        className="visually-hidden"
      >
        error
      </span>
      <div
        className="panel__body"
      >
        <ol
          className="list"
        >
          <li
            className="list__item "
          >
            <a
              className="list__link js-inpagelink"
              href="#container-test-number"
            >
              Enter a number
            </a>
          </li>
          <li
            className="list__item "
          >
            <a
              className="list__link js-inpagelink"
              href="#container-test-percent"
            >
              Enter a number less than or equal to 100
            </a>
          </li>
        </ol>
      </div>
    </div>
  </div>
</div>
`;

exports[`error panel with a regular heading and an input field 1`] = `
<div
  className="error"
>
  <div
    className=""
  >
    <div
      className="panel panel--error panel--no-title"
    >
      <span
        className="visually-hidden"
      >
        error
      </span>
      <div
        className="field"
      >
        <label
          className="label  "
          htmlFor="number"
        >
          Number of employees paid monthly
        </label>
        <input
          className="input input--text input-type__input  input--props.type"
          id="number"
          inputMode="numeric"
          pattern="[0-9]"
          type="text"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`error type panel with a regular heading and body 1`] = `
<div
  className="error"
>
  <div
    className=""
  >
    <div
      className="panel panel--error panel--no-title"
    >
      <span
        className="visually-hidden"
      >
        error
      </span>
      <div
        className="panel__body"
      >
        <p
          className="panel__error"
        >
          <strong
            className="font-weight--600"
          >
            Fix the following: 
          </strong>
        </p>
        Email address or password are incorrect
      </div>
    </div>
  </div>
</div>
`;

exports[`information panel with no heading and raw string for body 1`] = `
<div
  className="information"
>
  <div
    className=""
  >
    <div
      className="panel panel--information panel--no-title"
    >
      <span
        className="visually-hidden"
      >
        information
      </span>
      <div
        className="panel__body"
      >
        In principle but likewise practice this is a relatively simple but important message. The fact of the matter is that you had absolutely no idea and therefore did not know whether or not it was vital for you to spend the time to read this message until you actually started to consume the words which this message contains. Now that you have started to read it this message can help you realise that the nature of this message is to inform you that you have indeed actually now read this message.
      </div>
    </div>
  </div>
</div>
`;

exports[`success panel with a tick icon and a raw text body 1`] = `
<div
  className="success"
>
  <div
    className=""
  >
    <div
      className="panel panel--success panel--no-title"
    >
      <span
        className="panel__icon"
      >
        <svg
          className="svg-icon"
          focusable="false"
          viewBox="0 0 13 10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14.35,3.9l-.71-.71a.5.5,0,0,0-.71,0h0L5.79,10.34,3.07,7.61a.51.51,0,0,0-.71,0l-.71.71a.51.51,0,0,0,0,.71l3.78,3.78a.5.5,0,0,0,.71,0h0L14.35,4.6A.5.5,0,0,0,14.35,3.9Z"
            transform="translate(-1.51 -3.04)"
          />
        </svg>
      </span>
      <span
        className="visually-hidden"
      >
        success
      </span>
      <div
        className="panel__body"
      >
        Information has been successfully submitted
      </div>
    </div>
  </div>
</div>
`;

exports[`warning panel with an exclamation icon and raw string body 1`] = `
<div
  className="warning"
>
  <div
    className=""
  >
    <div
      className="panel panel--warning panel--no-title"
    >
      <span
        aria-hidden="true"
        className="panel__icon"
      >
        !
      </span>
      <span
        className="visually-hidden"
      >
        warning
      </span>
      <div
        className="panel__body"
      >
        All the information about this person will be deleted
      </div>
    </div>
  </div>
</div>
`;
